### ISSUE TITLE
how to clean memory?

### ISSUE THREAD

**Issue by FULLEST** [ultralytics/ultralytics#19849]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am using ultralytics in Celery, but I found that the memory usage does not disappear after the worker is destroyed. So I wrote a test code to verify this phenomenon. My code is as follows:
`import gc
import time

import torch
from ultralytics import YOLO

import os

os.environ["CUDA_VISIBLE_DEVICES"] = "0"

model = YOLO("media/weights/5fb7d6d0-c745-4969-bc7d-fac194d8077d.pt")

result = model.predict("media/images/frames/0a94bf04-f89b-41f3-a778-5bf0137eb869/1.jpg", stream=True)

time.sleep(10)

print("clean")

del model
del result
gc.collect()
torch.cuda.empty_cache()

for i in range(10000):
    torch.cuda.empty_cache()
    time.sleep(0.1)`
I ran this code and continuously observed the usage of video memory, but it did not release the memory occupied during the prediction process until the end of the run. May I ask how I should solve this problem?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @FULLEST, thank you for reaching out and for using Ultralytics 🚀! We appreciate the detailed explanation and code snippet you provided—it’s very helpful in understanding your concern.

If this is a 🐛 Bug Report, could you please ensure that your issue is reproducible and provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/)? This will help us identify and resolve the issue more effectively.

In the meantime, here are some suggestions to help you troubleshoot:  
1. Ensure you are using the latest version of the `ultralytics` package and all dependencies. Upgrade with:
```bash
pip install -U ultralytics
```
2. Verify that your environment meets the [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), including [Python>=3.8](https://www.python.org/) and [PyTorch>=1.8](https://pytorch.org/get-started/locally/).
3. If you're running YOLO in a Celery environment, consider isolating the issue by testing in a simpler setup to check if the behavior persists.

For additional support, you can explore the following resources:
- Check out the [Docs](https://docs.ultralytics.com/), especially the [Python Usage Guide](https://docs.ultralytics.com/usage/python/) for tips on managing resources.
- Join the Ultralytics community for discussions and advice:
  - [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat.
  - [Discourse](https://community.ultralytics.com/) for in-depth discussions.
  - [Ultralytics Subreddit](https://reddit.com/r/Ultralytics) for community insights.

## Verified Environments

For best results, consider running YOLO in one of these verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
- **Amazon** Deep Learning AMI: See the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker Image**: See the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

## Status

You can check the current status of our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests here:  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  
If the badge is green, all tests are passing across macOS, Windows, and Ubuntu.

This is an automated response, but no worries—an Ultralytics engineer will review your issue and assist you soon. Thanks again for your patience and for being part of the Ultralytics community! 😊

**Comment by FULLEST** [comment]
> You can try moving the model and results to CPU before deleting

thank you for your answer, i have already tried `model.to("cpu")` before `del model`, and it looks effective, but when i tried `result.cpu()` before `del result`, the memory has not changed, after using the method you provided, the final memory usage decreased from 484MB to 366MB. here is the new code:
```
import gc
import time

import torch
from ultralytics import YOLO

import os

os.environ["CUDA_VISIBLE_DEVICES"] = "0"

model = YOLO("media/weights/5fb7d6d0-c745-4969-bc7d-fac194d8077d.pt")

result = model.predict("media/images/frames/0a94bf04-f89b-41f3-a778-5bf0137eb869/1.jpg")[0]

time.sleep(10)

print("clean")

model.to("cpu")
for r in result:
    r.cpu()
del model
del result
gc.collect()
torch.cuda.empty_cache()

for i in range(10000):
    torch.cuda.empty_cache()
    time.sleep(0.1)
```

**Comment by Y-T-G** [comment]
Is this GPU memory usage?

**Comment by FULLEST** [comment]
> Is this GPU memory usage?

yes

**Comment by edargham** [comment]
`model.predictor.dataset.close()`
It seems when `stream=True` the `LoadStreams` pytorch dataloader used for generating inputs isn't being properly closed after the worker is destroyed.

This line was able to clear most of the leaked memory in my case, but it seems there's no proper cleanup for the models after workers are destroyed.

**Comment by glenn-jocher** [comment]
@FULLEST Thank you for your detailed explanation and test case. The issue you're encountering is related to how YOLO handles resources during streaming prediction.

When using `stream=True`, the `LoadStreams` dataloader isn't being properly closed. @edargham's suggestion to call `model.predictor.dataset.close()` is correct and should help release most of the resources.

Additionally, for proper cleanup with results, you need to apply `.cpu()` correctly. The Results object methods return new objects rather than modifying in-place, so your loop should be:

```python
result = result.cpu()  # For a single result
# Or for multiple results:
results = [r.cpu() for r in results]
```

Here's a more complete cleanup approach:

```python
# After prediction
model.to("cpu")
result = result.cpu()  # Correctly move result to CPU
model.predictor.dataset.close()  # Close the dataset if using stream=True
del model
del result
gc.collect()
torch.cuda.empty_cache()
```

This is a known issue with resource management, particularly in worker environments like Celery. We appreciate you bringing this to our attention and will look into improving the cleanup process.
